balancer/catabalancer/catalyst_balancer.go

```diff
@@ -291,7 +327,58 @@ func truncateReturned(scoredNodes []ScoredNode, numNodes int) []ScoredNode {
 	return scoredNodes[:numNodes]
 }
 
+func (c *CataBalancer) RefreshNodes() error {
+	log.LogNoRequestID("catabalancer refreshing nodes")
+	if c.MetricsDB == nil {
+		return fmt.Errorf("node stats DB was nil")
+	}
+
+	query := "SELECT stats FROM node_stats"
+	rows, err := c.MetricsDB.Query(query)
+	if err != nil {
+		return fmt.Errorf("failed to query node stats: %w", err)
+	}
+	defer rows.Close()
+
+	// Process the result set
+	for rows.Next() {
+		var statsBytes []byte
+		if err := rows.Scan(&statsBytes); err != nil {
+			return fmt.Errorf("failed to scan node stats row: %w", err)
+		}
+
+		var event NodeUpdateEvent
+		err = json.Unmarshal(statsBytes, &event)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal node update event: %w", err)
+		}
+
+		if isStale(event.NodeMetrics.Timestamp, c.metricTimeout) {
+			log.LogNoRequestID("catabalancer skipping stale data while refreshing", "nodeID", event.NodeID, "timestamp", event.NodeMetrics.Timestamp)
+			continue
+		}
+
+		c.UpdateNodes(event.NodeID, event.NodeMetrics)
```

Comment:
Do we actually need to keep any internal state for nodes? If we make a query to DB with every playback redirect request, then couldn't we reason everything from there?

Reply:
Yep sorry I totally forgot to mention refactoring that as a plan for another PR, in two minds about whether I should do that in this PR but I think since it's been approved I think I'll just do it now but in a fresh PR.